Sources/StreamChat/Workers/ChannelListCleanupUpdater.swift

```diff
@@ -0,0 +1,110 @@
+//
+// Copyright © 2021 Stream.io Inc. All rights reserved.
+//
+
+import CoreData
+
+class ChannelListCleanupUpdater<ExtraData: ExtraDataTypes>: Worker {
+    // MARK: - Properties
+    
+    private let channelListUpdater: ChannelListUpdater<ExtraData>
+    
+    // MARK: - Init
+    
+    override init(
+        database: DatabaseContainer,
+        apiClient: APIClient
+    ) {
+        channelListUpdater = ChannelListUpdater(database: database, apiClient: apiClient)
+        super.init(
+            database: database,
+            apiClient: apiClient
+        )
+    }
+    
+    init(
+        database: DatabaseContainer,
+        apiClient: APIClient,
+        channelListUpdater: ChannelListUpdater<ExtraData>
+    ) {
+        self.channelListUpdater = channelListUpdater
+        super.init(
+            database: database,
+            apiClient: apiClient
+        )
+    }
+    
+    func cleanupChannelList() {
+        database.write { _ in
+            self.allChannels.forEach { $0.syncFailedCleanUp() }
+        } completion: { error in
+            if let error = error {
+                log.error("Failed cleaning up channels data: \(error).")
+            }
+            self.updateChannels()
+        }
+    }
+    
+    // MARK: - Private
+    
+    private func updateChannels() {
+        database.backgroundReadOnlyContext.perform {
+            self.queries.forEach {
+                self.channelListUpdater.update(channelListQuery: $0) { error in
+                    if let error = error {
+                        log.error("Internal error. Failed to update ChannelListQueries for the new channel: \(error)")
+                    }
+                }
+            }
```

Comment:
Same here, `self.queries` is not the best way to handle this when it's possible to access it only from within the `backgroundReadOnlyContext` queue.